feature/path entropy

NicoloPosta opened the pull request:
- Iniziato metodo Path Entropy
- Terminato path_entropy

NicoloPosta committed:
Iniziato metodo Path Entropy

Da aggiustare il ritorno della funzione, capire se lo weight è giusto e se le formule sono state interpretate bene

Co-Authored-By: Tommaso Romani <44830726+TommasoRomani@users.noreply.github.com>

NicoloPosta committed:
Terminato path_entropy

Aggiornamento e refactoring della funzione.

ncvescera commented on src/social_network_link_prediction/other_methods/information_theory/path_entropy.py:
```diff
@@ -0,0 +1,45 @@
+            tmp += (1 / (i - 1)) * simple_path_entropy(paths = paths, G = G)
+        tmp = tmp - new_link_entropy(G, elem[0], elem[1])
+        similarity_matrix[elem[0],elem[1]] = tmp
+    return similarity_matrix
```
Deve ritornare un `csr_matrix`

```suggestion
def path_entropy(G:nx.Graph, max_path:int = 3) -> csr_matrix:

    similarity_matrix = lil_matrix((G.number_of_nodes(), G.number_of_nodes()))

    missing_edges = list(nx.complement(G).edges())

    for elem in missing_edges:
        paths = nx.all_simple_paths(G, elem[0], elem[1], max_path)
        tmp = 0
        for i in range(2,(max_path+1)):
            tmp += (1 / (i - 1)) * simple_path_entropy(paths = paths, G = G)
        tmp = tmp - new_link_entropy(G, elem[0], elem[1])
        similarity_matrix[elem[0],elem[1]] = tmp
    return similarity_matrix.tocsr()
```

ncvescera replied:
Risolto in 5b4ab4d

ncvescera commented on src/social_network_link_prediction/other_methods/information_theory/path_entropy.py:
```diff
@@ -0,0 +1,45 @@
+
+# Calcola l'entropia basata sulla probabilità a priori della creazione del link diretto
+# tra le coppie di noti senza link diretti
+def new_link_entropy(G:nx.Graph, a:int, b:int) -> int:
```
`a` e `b` non sono interi ma etichette del grafo

```suggestion
def new_link_entropy(G:nx.Graph, a, b) -> int:
```

ncvescera replied:
Risolto in 5b4ab4d

ncvescera commented on src/social_network_link_prediction/other_methods/information_theory/path_entropy.py:
```diff
@@ -0,0 +1,45 @@
+import math
+from scipy.sparse import lil_matrix
+
+def path_entropy(G:nx.Graph, max_path:int = 3) -> lil_matrix:
```
Si rompe con grafi che hanno come etichette delle stringhe

ncvescera replied:
Risolto in 5b4ab4d

ncvescera committed:
Path Entropy Fix

Risolti piccoli alcuni piccoli problemi e adattato codice a
gestire grafi aventi nodi con etichette stringhe.
